File: netlifyfunctions/functions/signup.js
const { Client } = require('pg');
const bcrypt = require('bcryptjs');

// Helper headers for JSON responses
const HEADERS = {
    'Content-Type': 'application/json',
    'Access-Control-Allow-Origin': '*',
};

exports.handler = async (event) => {
    if (event.httpMethod !== 'POST') {
        return {
            statusCode: 405,
            headers: HEADERS,
            body: JSON.stringify({ error: 'Method not allowed' })
        };
    }

    // --- FIX APPLIED HERE ---
    // Now using DATABASE_URL to match the Netlify environment variable key
    const client = new Client({
        connectionString: process.env.DATABASE_URL, 
        ssl: { rejectUnauthorized: false }
    });

    try {
        const { name, email, password, role } = JSON.parse(event.body);

        if (!name || !email || !password || !role) {
            return {
                statusCode: 400,
                headers: HEADERS,
                body: JSON.stringify({ error: 'Missing required fields: name, email, password, role' })
            };
        }

        await client.connect();

        // 1. Check if user already exists (NeonDB: findOne)
        const checkSql = `SELECT id FROM users WHERE email = $1;`;
        const existingUserResult = await client.query(checkSql, [email]);

        if (existingUserResult.rows.length > 0) {
            // User already exists
            return {
                statusCode: 409, // Conflict
                headers: HEADERS,
                body: JSON.stringify({ error: 'User already exists with this email.' }),
            };
        }

        // 2. Hash password
        const hashedPassword = await bcrypt.hash(password, 10);

        // 3. Insert new user (NeonDB: insertOne)
        const insertSql = `
            INSERT INTO users (name, email, password_hash, role, balance, created_at)
            VALUES ($1, $2, $3, $4, $5, NOW())
            RETURNING id;
        `;
        const insertValues = [name, email, hashedPassword, role, 0]; // Initial balance is 0

        const result = await client.query(insertSql, insertValues);

        // PostgreSQL uses RETURNING to give back the inserted row data, specifically the 'id'.
        const userId = result.rows[0].id;

        // SUCCESS: Returns clean JSON
        return {
            statusCode: 201,
            headers: HEADERS,
            body: JSON.stringify({
                status: 'success',
                message: 'Account created successfully',
                user_id: userId,
            }),
        };

    } catch (error) {
        console.error('Signup Handler Error:', error.message);
        return {
            statusCode: 500,
            headers: HEADERS,
            body: JSON.stringify({
                error: 'Internal server error.',
                details: error.message
            }),
        };
    } finally {
        // IMPORTANT: Ensure the connection is closed after the operation
        if (client) {
            await client.end().catch(e => console.error("Error closing PG client:", e));
        }
    }
};

File: netlifyfunctions/functions/signup.py
import json
import os
import psycopg2
import hashlib
import re

def handler(event, context):
    """
    Netlify Python Function: /signup
    Handles new user registration for Vanguard Escrow.
    """

    # Parse request body
    try:
        data = json.loads(event.get("body", "{}"))
        name = data.get("name", "").strip()
        email = data.get("email", "").strip().lower()
        password = data.get("password", "")
        role = data.get("role", "").strip().lower()
    except Exception as e:
        return {
            "statusCode": 400,
            "body": json.dumps({"error": "Invalid JSON body", "details": str(e)})
        }

    # Simple validation
    if not name or not email or not password or not role:
        return {
            "statusCode": 400,
            "body": json.dumps({"error": "All fields are required"})
        }

    if role not in ["buyer", "seller"]:
        return {
            "statusCode": 400,
            "body": json.dumps({"error": "Invalid role. Must be buyer or seller"})
        }

    if not re.match(r"[^@]+@[^@]+\.[^@]+", email):
        return {
            "statusCode": 400,
            "body": json.dumps({"error": "Invalid email format"})
        }

    # Connect to Neon database
    try:
        conn = psycopg2.connect(
            host=os.getenv("DB_HOST"),
            database=os.getenv("DB_NAME"),
            user=os.getenv("DB_USER"),
            password=os.getenv("DB_PASS"),
            sslmode="require"
        )
        cur = conn.cursor()
    except Exception as e:
        return {
            "statusCode": 500,
            "body": json.dumps({"error": "Database connection failed", "details": str(e)})
        }

    # Check if email already exists
    try:
        cur.execute("SELECT id FROM users WHERE email = %s;", (email,))
        if cur.fetchone():
            cur.close()
            conn.close()
            return {
                "statusCode": 409,
                "body": json.dumps({"error": "Email already registered"})
            }
    except Exception as e:
        cur.close()
        conn.close()
        return {
            "statusCode": 500,
            "body": json.dumps({"error": "Error checking user existence", "details": str(e)})
        }

    # Hash password (SHA256)
    password_hash = hashlib.sha256(password.encode()).hexdigest()

    # Insert new user
    try:
        cur.execute(
            """
            INSERT INTO users (name, email, password_hash, role)
            VALUES (%s, %s, %s, %s)
            RETURNING id;
            """,
            (name, email, password_hash, role)
        )
        user_id = cur.fetchone()[0]
        conn.commit()
        cur.close()
        conn.close()
    except Exception as e:
        cur.close()
        conn.close()
        return {
            "statusCode": 500,
            "body": json.dumps({"error": "Database insert failed", "details": str(e)})
        }

    # Success response
    return {
        "statusCode": 200,
        "headers": {"Content-Type": "application/json"},
        "body": json.dumps({
            "message": "Signup successful",
            "user_id": user_id,
            "role": role
        })
    }
